pkg/serverless/logs/logs_collector.go

```diff
@@ -0,0 +1,216 @@
+		extraTags:              extraTags,
+		logsEnabled:            logsEnabled,
+		enhancedMetricsEnabled: enhancedMetricsEnabled,
+		executionContext:       executionContext,
+		handleRuntimeDone:      handleRuntimeDone,
+		process_once:           &sync.Once{},
+	}
+}
+
+// Start processing logs. Can be called multiple times, but only the first invocation will be effective.
+func (lc *LambdaLogsCollector) Start() {
+
+	lc.process_once.Do(func() {
+		// After a timeout, there may be queued logs that will be immediately sent to the logs API.
+		// We want to use the restored execution context for those logs.
+		state := lc.executionContext.GetCurrentState()
+
+		log.Debugf("Starting Log Collection with ARN: %s and RequestId: %s", state.ARN, state.LastLogRequestID)
+
+		// Once we have an ARN, we can start processing logs
+		lc.arn = state.ARN
+		lc.lastRequestID = state.LastRequestID
+		lc.coldstartRequestID = state.ColdstartRequestID
+		lc.invocationStartTime = state.StartTime
+		lc.invocationEndTime = state.EndTime
+
+		go func() {
+			for messages := range lc.In {
+				lc.processLogMessages(messages)
+			}
+		}()
+	})
+}
+
+// Shutdown the log collector
+func (lc *LambdaLogsCollector) Shutdown() {
+	close(lc.In)
+}
+
+// shouldProcessLog returns whether or not the log should be further processed.
+func shouldProcessLog(message *LambdaLogAPIMessage) bool {
+	// Making sure that empty logs are not uselessly sent
+	if len(message.stringRecord) == 0 && len(message.objectRecord.requestID) == 0 {
+		return false
+	}
+
+	return true
+}
+
+func createStringRecordForReportLog(startTime, endTime time.Time, l *LambdaLogAPIMessage) string {
+	runtimeDurationMs := float64(endTime.Sub(startTime).Milliseconds())
+	postRuntimeDurationMs := l.objectRecord.reportLogItem.durationMs - runtimeDurationMs
+	stringRecord := fmt.Sprintf("REPORT RequestId: %s\tDuration: %.2f ms\tRuntime Duration: %.2f ms\tPost Runtime Duration: %.2f ms\tBilled Duration: %d ms\tMemory Size: %d MB\tMax Memory Used: %d MB",
+		l.objectRecord.requestID,
+		l.objectRecord.reportLogItem.durationMs,
+		runtimeDurationMs,
+		postRuntimeDurationMs,
+		l.objectRecord.reportLogItem.billedDurationMs,
+		l.objectRecord.reportLogItem.memorySizeMB,
+		l.objectRecord.reportLogItem.maxMemoryUsedMB,
+	)
+	initDurationMs := l.objectRecord.reportLogItem.initDurationMs
+	if initDurationMs > 0 {
+		stringRecord = stringRecord + fmt.Sprintf("\tInit Duration: %.2f ms", initDurationMs)
+	}
+
+	return stringRecord
+}
+
+// removeInvalidTracingItem is a temporary fix to handle malformed JSON tracing object
+func removeInvalidTracingItem(data []byte) []byte {
+	return []byte(strings.ReplaceAll(string(data), ",\"tracing\":}", ""))
+}
+
+func (lc *LambdaLogsCollector) processLogMessages(messages []LambdaLogAPIMessage) {
+	// sort messages by time (all from the same time zone) in ascending order.
+	sort.Slice(messages, func(i, j int) bool {
+		return messages[i].time.Before(messages[j].time)
+	})
```

Comment:
Will we still need this sort with your changes?

Reply:
I don't think it hurts to leave it here, just in case the API ever returns things out of order. We depend on the logs being sorted in our logic, so we should ensure they are.